From 5861fd46873d9b46c03e67f11c7d7fffabcab1ab Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Fri, 12 Aug 2022 22:18:44 +0300
Subject: [PATCH] imx8mp: ddr: Pass the dram info from SPL to U-Boot

Some DRAM vendors report the same dram information in mr[5-9] register
for the different dram sizes. As a result the U-Boot does not have an
option to get the dram size from the mr[5-8] registers. In order to
solve this problem a SHARED_DDR_INFO approach was implemented

SHARED_DDR_INFO is the area that stores the dram information detected by
the SPL code. The same address is accessible from the U-Boot code. As of now
SHARED_DDR_INFO address is used for handing over the memory info to the
U-Boot.

SHARED_DDR_INFO address is equal to the SPL_STACK address.
NOTE: the stack grows down and it never goes back to is bottom at SPL.

TBD:
This approach will be removed as soon as the handoff code gets provided
for the platform.

Signed-off-by: Valentin Raevsky <valentin@compulab.co.il>
---
 board/compulab/plat/imx8mp/ddr/ddr.c        | 8 ++++++++
 board/compulab/plat/imx8mp/ddr/ddr_common.c | 7 +++++++
 include/configs/compulab-imx8m-plus.h       | 4 +++-
 3 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/board/compulab/plat/imx8mp/ddr/ddr.c b/board/compulab/plat/imx8mp/ddr/ddr.c
index 76dae8d9ef..daee329f73 100644
--- a/board/compulab/plat/imx8mp/ddr/ddr.c
+++ b/board/compulab/plat/imx8mp/ddr/ddr.c
@@ -143,6 +143,13 @@ static inline void lpddr4_data_set(struct lpddr4_tcm_desc *lpddr4_tcm_desc) {
 	cl_eeprom_buffer_write(0, (uchar *)lpddr4_tcm_desc, sizeof(struct lpddr4_tcm_desc));
 }
 
+static inline void spl_dram_share_info(void) {
+#ifdef SHARED_DDR_INFO
+    struct lpddr4_tcm_desc *lpddr4_tcm_desc = (void *) SHARED_DDR_INFO;
+    memcpy(lpddr4_tcm_desc, SPL_TCM_DATA, sizeof(struct lpddr4_tcm_desc));
+#endif
+}
+
 void spl_dram_init(void)
 {
     int rc=0;
@@ -153,4 +160,5 @@ void spl_dram_init(void)
         printf("%s Reset ... \n",__func__);
         do_reset_spl();
     }
+    spl_dram_share_info();
 }
diff --git a/board/compulab/plat/imx8mp/ddr/ddr_common.c b/board/compulab/plat/imx8mp/ddr/ddr_common.c
index 520ccd3ee4..85c9b4fba6 100644
--- a/board/compulab/plat/imx8mp/ddr/ddr_common.c
+++ b/board/compulab/plat/imx8mp/ddr/ddr_common.c
@@ -60,6 +60,12 @@ const struct lpddr4_desc *lpddr4_get_desc_by_id(unsigned int id) {
 	return NULL;
 }
 
+#ifdef SHARED_DDR_INFO
+size_t lppdr4_get_ramsize() {
+	struct lpddr4_tcm_desc *lpddr4_tcm_desc = (void *) SHARED_DDR_INFO;
+	return lpddr4_tcm_desc->size;
+}
+#else
 size_t lppdr4_get_ramsize() {
 	size_t ramsize = 0;
 	unsigned int id = lpddr4_get_mr();
@@ -68,3 +74,4 @@ size_t lppdr4_get_ramsize() {
 		ramsize = desc->size;
 	return ramsize;
 }
+#endif
diff --git a/include/configs/compulab-imx8m-plus.h b/include/configs/compulab-imx8m-plus.h
index 573b0e3370..1066c7c10f 100644
--- a/include/configs/compulab-imx8m-plus.h
+++ b/include/configs/compulab-imx8m-plus.h
@@ -19,8 +19,10 @@
 #define CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR	0x300
 #define CONFIG_SYS_UBOOT_BASE	(QSPI0_AMBA_BASE + CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_SECTOR * 512)
 
+#define SPL_STACK_ADDR			0x96dff0
+#define SHARED_DDR_INFO			SPL_STACK_ADDR
 #ifdef CONFIG_SPL_BUILD
-#define CONFIG_SPL_STACK		0x96dff0
+#define CONFIG_SPL_STACK		SPL_STACK_ADDR
 #define CONFIG_SPL_BSS_START_ADDR	0x96e000
 #define CONFIG_SPL_BSS_MAX_SIZE		SZ_8K	/* 8 KB */
 #define CONFIG_SYS_SPL_MALLOC_START	0x42200000
-- 
2.17.1

